core/domain/draft_upgrade_services.py

```diff
@@ -87,6 +87,70 @@ def try_upgrading_draft_to_exp_version(
 class DraftUpgradeUtil(python_utils.OBJECT):
     """Wrapper class that contains util functions to upgrade drafts."""
 
+    @classmethod
+    def _convert_states_v30_dict_to_v31_dict(cls, draft_change_list):
+        """Converts draft change list from state version 30 to 31. State
+        Version 31 adds the duration float for the Voiceover section of state.
+
+        Args:
+            draft_change_list: list(ExplorationChange). The list of
+                ExplorationChange domain objects to upgrade.
+
+        Returns:
+            list(ExplorationChange). The converted draft_change_list.
+        """
+        for i, change in enumerate(draft_change_list):
+            if (change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY and
+                    change.property_name ==
+                    exp_domain.STATE_PROPERTY_RECORDED_VOICEOVERS):
+                # Get the language code to access the language code correctly.
+                new_change = change.new_value['voiceovers_mapping']
+                language_code = new_change['content'].keys()
+                # Initialize the value to migrate draft state to v31.
+                new_change['content'][language_code[0]]['duration'] = 0
+                draft_change_list[i] = exp_domain.ExplorationChange({
+                    'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                    'property_name': (
+                        exp_domain.STATE_PROPERTY_RECORDED_VOICEOVERS),
+                    'state_name': change.state_name,
+                    'new_value': {
+                        'voiceovers_mapping': new_change
+                    }
+                })
+        return draft_change_list
+
+
+    @classmethod
+    def _convert_states_v29_dict_to_v30_dict(cls, draft_change_list):
+        """Converts draft change list from state version 29 to 30. State
+        version 30 replaces tagged_misconception_id with
+        tagged_skill_misconception_id.
+
+        Args:
+            draft_change_list: list(ExplorationChange). The list of
+                ExplorationChange domain objects to upgrade.
+
+        Returns:
+            list(ExplorationChange). The converted draft_change_list.
+        """
+        for i, change in enumerate(draft_change_list):
+            if (change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY and
```

Comment:
Can you please ask the contributor who has written the v30 migration to review this?

Reply:
okay I will reach out to that contributor.

Reply:
Hey @sophiewu6 just looking for your review for this. Would be grateful for your guidance on this.

Reply:
@mzaman07 Hi, I would like to help, but I am a little confused about why you implemented another _convert_states_v29_dict_to_v30_dict method. If you're trying to upgrade it once again, you could do _convert_states_v30_dict_to_v31_dict instead.

Reply:
Whoops I'll get rid of it. My bad.

Reply:
Now it should be good @sophiewu6.

Reply:
Finally got this working for drafts. Done.